### PR TITLE
 Refactor | Implement main and admin route handlers with comprehensive test suites and CI integration | Fixed a lot of Sonar Issues | Part 1

### PR DESCRIPTION
**Pull request by Debagnik**
## Summary by CodeRabbit

* **Bug Fixes**
  * Hardened input validation, sanitization, and error handling across registration, posts, comments, and admin flows
  * Fixed environment-variable parsing and several control-flow issues

* **Chores**
  * Updated CI workflow and SonarCloud/SonarLint configuration; added Node setup, dependency install, and coverage reporting before scans
  * Added TypeScript node type definitions to devDependencies

* **Tests**
  * Improved test cleanup and expanded/adjusted test suites to cover new edge cases and deprecated routes

* **Documentation**
  * Updated API docs to use a unique identifier for post lookup

### CODE REVIEW

**Comment by Debagnik** [comment]
@Coderabbitai pause review

File: server/routes/admin.js [hunk 1]
```diff
@@ -1307,88 +1292,75 @@ router.post('/edit-site-config', authToken, genericAdminRateLimiter, async (req,
...
+    const paginationLimit = Number.parseInt(req.body.defaultPaginationLimit, 10) || 10;
+    if (Number.isNaN(paginationLimit) || paginationLimit < 1 || paginationLimit > 100) {
+      console.error('Invalid pagination limit');
+      throw new Error('Webmaster tried updating invalid pagination limit');
+    }
+
+    const searchLimit = Number.parseInt(req.body.searchLimit, 10) || 10;
+    if (Number.isNaN(searchLimit) || searchLimit < 1 || searchLimit > 50) {
+      console.error('Invalid search limit');
+      throw new Error('Webmaster tried updating invalid search limit');
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: @CoderabbitAi that means that isNaN() check isn't required anymore, cause then it will always fallback to a number. and if req.body.searchLimit = 0 then the parseint() will still return 0 and it will not fallback and then it will throw a validation error on searchLimit < 1 which is correct. I don't want 0 or negative numbers

File: server/routes/admin.js [hunk 2]
```diff
@@ -1456,19 +1442,19 @@
...
     if (!isUserIdValid) {
-      console.error({ code: 404, Status: 'Not found', message: `userID: ${req.params.id} is not valid` });
+      console.error({ code: 404, Status: 'Not found', message: `userID: ${sanitizeLogParams(req.params.id)} is not valid` });
       throw new Error('User ID is invalid');
     }
     const userToDelete = await user.findById(req.params.id);
     if (!userToDelete) {
-      console.warn('User not found', req.params.id);
-      throw new Error('Current user not found');
+      console.error('User not found', sanitizeLogParams(req.params.id));
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: @coderabbitai this comment seems BS cause the regex /[\r\n]/g is same as CONSTANTS.LOGS_ESCAPE_REGEX,

is this a false positive?

File: Documentation/Project-walnut/main/getPostById.bru [hunk 3]
```diff
@@ -34,17 +34,17 @@ headers {
...
-  Displays a specific blog post based on its MongoDB _id. Includes post metadata, author details, comments, and conditionally shows the post based on approval status or user privileges.
+  Displays a specific blog post based on its unique Id generated based on title. Includes post metadata, author details, comments, and conditionally shows the post based on approval status or user privileges.
   
   Middleware:
   * fetchSiteConfig: Attaches site-wide config to res.locals.
   * csrfProtection: Adds CSRF protection to the rendered post page.
   
   Parameters:
-  * :id (Path Param): MongoDB _id of the blog post to retrieve.
+  * :uniqueId (Path Param): Unique Id of the blog post to retrieve.
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply Documentation/Project-walnut/main/getPostById.bru:46]: this is also fixed

File: server/models/comments.js [hunk 4]
```diff
@@ -8,7 +8,7 @@
         minlength: [3, 'Commenter name must be at least 3 characters!'],
         maxlength: [50, 'Commenter name is too long!'],
         trim: true,
-        set: value => value.replace(/<\/?[^>]+(>|$)/g, '')
+        set: value => value.replaceAll(/<\/?[^>]+>?/g, '')
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/models/comments.js]: fixed, used sanitise HTMl instead

File: server/models/comments.js [hunk 5]
```diff
@@ -17,7 +17,7 @@
         minlength: [1, 'Comment must be at least 1 character!'],
         maxlength: [500, 'Comment is too long!'],
         trim: true,
-        set: value => value.replace(/<\/?[^>]+(>|$)/g, '')
+        set: value => value.replaceAll(/<\/?[^>]+>?/g, '')
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/models/comments.js]: used SanitiseHtml instead

File: server/routes/admin.js [hunk 6]
```diff
@@ -198,33 +202,35 @@
...
     if (!usernameRegex.test(username)) {
       const env = process.env.NODE_ENV;
-      if (env && env.toLowerCase() === "production") {
-        console.error(400, 'Invalid username format');
-      } else {
-        console.error(400, 'Invalid username format', username);
-      }
+      if (env?.toLowerCase() === "production") {
+      logger.error(400, 'Invalid username format');
+    } else {
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed

File: server/routes/admin.js [hunk 7]
```diff
@@ -198,33 +202,35 @@
...
+      logger.error(400, 'Invalid username format', username);
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: Fixed

File: server/routes/admin.js [hunk 8]
```diff
@@ -311,38 +317,38 @@
...
-      throw Error('Chica, Its not this easy to dupe me, Try harder');
+    if (typeof username !== 'string') {
+      throw new TypeError('Chica, Its not this easy to dupe me, Try harder');
     }
 
     //checks if the user exists
     const currentUser = await user.findOne({ username: { $eq: username } });
     if (!currentUser) {
-      console.error('invalid Username for user: ', username);
+      logger.error(`invalid Username for user: ${username}`);
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed

File: server/routes/admin.js [hunk 9]
```diff
@@ -311,38 +317,38 @@
...
       throw new Error('Either username or password dont match, Invalid credentials');
     }
 
     if (currentUser.isPasswordReset) {
-      console.error('Login Disabled for this Username:', username);
+      logger.error(`Login Disabled for this Username: ${username}`);
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed

File: server/routes/admin.js [hunk 10]
```diff
@@ -311,38 +317,38 @@
...
       throw new Error('Login Disabled for this Username, Contact Webmaster');
     }
 
     //password validity check
     const isPasswordValid = await bcrypt.compare(password, currentUser.password);
     if (!isPasswordValid) {
-      console.error('invalid password for user: ', username);
+      logger.error(`invalid password for user: ${username}`);
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js:346]: Fixed

File: server/routes/admin.js [hunk 11]
```diff
@@ -1307,88 +1292,75 @@
...
+      homepageWelcomeImage: validHomePageImageUri,
+      copyrightText: sanitizeHtml(String(req.body.copyrightText), CONSTANTS.SANITIZE_FILTER),
+      searchLimit: searchLimit,
+      cloudflareSiteKey: sanitizeHtml(String(req.body.cloudflareSiteKey), CONSTANTS.SANITIZE_FILTER),
+      cloudflareServerKey: sanitizeHtml(String(req.body.cloudflareServerKey), CONSTANTS.SANITIZE_FILTER),
+      siteVisitCounter: siteVisitCounter
+    };
+
+    if (globalSiteConfig) {
+      await siteConfig.findOneAndUpdate({}, configData, { new: true });
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed

File: server/routes/admin.js [hunk 12]
```diff
@@ -1455,25 +1430,25 @@
...
     if (!isUserIdValid) {
-      console.error({ code: 404, Status: 'Not found', message: `userID: ${req.params.id} is not valid` });
+      logger.error({ code: 404, Status: 'Not found', message: `userID: ${sanitizeLogParams(req.params.id)} is not valid` });
       throw new Error('User ID is invalid');
     }
     const userToDelete = await user.findById(req.params.id);
     if (!userToDelete) {
-      console.warn('User not found', req.params.id);
-      throw new Error('Current user not found');
+      logger.error('User not found', sanitizeLogParams(req.params.id));
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed

File: server/routes/admin.js [hunk 13]
```diff
@@ -1549,13 +1524,13 @@
...
     if (!isUserIdValid) {
-      console.error("Invalid username");
+      logger.error("Invalid username");
       throw new Error("User id to be edited is invalid");
     }
 
     const selectedUser = await user.findOne({ _id: req.params.id });
     if (!selectedUser) {
-      console.error('User not found', req.params.id);
+      logger.error('User not found', req.params.id);
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: Fixed

File: server/routes/admin.js [hunk 14]
```diff
@@ -1935,7 +1910,7 @@
     const { username, tempPassword, newPassword, confirmPassword } = req.body;
 
     if (!username || !tempPassword || !newPassword || !confirmPassword) {
-      console.log({ 'status': 400, 'message': 'one or more required feild missing' })
+      logger.warn({ 'status': 400, 'message': 'one or more required field missing' })
```
[the remark this answers is not in the document]
Comment (Debagnik) [reply server/routes/admin.js]: fixed
